astropy/io/fits/fitsrec.py

```diff
@@ -814,6 +816,8 @@ def _convert_p(self, column, field, recformat):
                 dt = np.dtype(recformat.dtype)
                 arr_len = count * dt.itemsize
                 dummy[idx] = raw_data[offset:offset + arr_len].view(dt)
+                if column.dim:
+                    dummy[idx] = dummy[idx].reshape(vla_shape)
```

Comment:
Here the VLA is reshaped consistently with TDIM instructions.

Reply:
I realized that this is working in this specific case but, as VLA have variable length, this may not work when different rows have different length.... I'm working on figure this out, and adding a related test too.